data_structures/treap.cpp

```diff
@@ -0,0 +1,234 @@
+
+/**
+ * @file
+ * @brief A balanced binary search tree (BST) on the basis of binary search tree
+ * and heap: the [Treap](https://en.wikipedia.org/wiki/Treap) algorithm
+ * implementation
+ *
+ * @details
+ * Implementation of the treap data structre
+ *
+ * Support operations including insert, erase, and query (the rank of specified
+ * element or the element ranked x) as the same as BST
+ *
+ * But these operations take O(log N) time, since treap keeps property of heap
+ * using rotate operation, and the desired depth of the tree is O(log N).
+ * There's very little chance that it will degenerate into a chain like BST
+ *
+ * @author [Kairao ZHENG](https://github.com/fgmn)
+ */
+
+#include <cassert>   /// For assert
+#include <iostream>  /// For IO operations
+
+const int maxNode = 1e6 + 5;  //< maximum number of nodes
+/**
+ * @brief Struct representation of the treap
+ */
+struct Treap {
+    int root;                //< root of the treap
```

Comment:
I believe this can be initialized. It is also a good practice to do that (don't forget the triple `/`).
```suggestion
    int root = 0;                ///< root of the treap
```